homeassistant/components/recorder/purge.py

```diff
@@ -23,6 +25,37 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+class PurgeRecoverableError(Exception):
+    """Recoverable exception (SQLAlchemyError, some OperationalError) and purge can be retryed."""
+
+
+@contextmanager
+def purge_session(instance: Recorder) -> Generator[Session, None, None]:
+    """Session manager that provides purge exception handling."""
+    try:
+        with session_scope(session=instance.get_session()) as session:  # type: ignore
+            yield session
+    except OperationalError as err:
+        # Retry when one of the following MySQL errors occurred:
+        # 1205: Lock wait timeout exceeded; try restarting transaction
+        # 1206: The total number of locks exceeds the lock table size
+        # 1213: Deadlock found when trying to get lock; try restarting transaction
+        if instance.engine.driver in ("mysqldb", "pymysql") and err.orig.args[0] in (
+            1205,
+            1206,
+            1213,
+        ):
+            _LOGGER.info("%s; purge not completed, retrying", err.orig.args[1])
```

Comment:
Any ideas on how to get `pytest` to trigger this branch for codecov ? `instance.engine.driver` is hardcoded to "sqlite" in test env and ignores config option `db_uri`.

Reply:
You could use `patch.object` on `hass.data[DATA_INSTANCE].engine` to mock the driver.

Reply:
The statistics PR (https://github.com/home-assistant/core/pull/49852) resolved this.